File: src/UnwindMC/Generation/Ast/Transformations/FixupZeroAssignment.cs
using static UnwindMC.Generation.Ast.PatternMatching;
using static UnwindMC.Util.PatternMatching;

namespace UnwindMC.Generation.Ast.Transformations
{
    public class FixupZeroAssignment : NodeTransformerBase
    {
        public override AssignmentNode Transform(AssignmentNode node)
        {
            var v = Capture<ValueNode>();
            return Match(node.Expression,
                Binary(Var(C(node.Var.Name)), C(Operator.And), v % Value(C(0)))
                    .Then(() => new AssignmentNode(node.Var, (ValueNode)v)),
                _.Then(() => node));
        }
    }
}

File: src/UnwindMC/Generation/Ast/Transformations/FixupZeroAssignment.fs
module FixupZeroAssignment

open Ast

let transformer: Transformer.T =
    {
        Transformer.def with
            transformStatement =
                fun t stmt ->
                    match stmt with
                    | Assignment (Var name, expr) ->
                        match expr with
                        | Binary (Operator.And, VarRef (Var v), Value 0) when name = v -> Assignment (Var v, Value 0)
                        | _ -> Transformer.def.transformStatement t stmt
                    | _ -> Transformer.def.transformStatement t stmt
    }
